mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,614 @@
+    init_x = (init_obs_batched, init_done)  # pack the RNN dummy inputs
+    init_hidden_state = ScannedRNN.initialize_carry(
+        (cfg.arch.num_envs, num_agents), cfg.network.hidden_state_dim
+    )  # (B, A, ...)
+
+    # Making recurrent Q network
+    pre_torso = hydra.utils.instantiate(cfg.network.q_network.pre_torso)
+    post_torso = hydra.utils.instantiate(cfg.network.q_network.post_torso)
+    q_net = RecQNetwork(
+        pre_torso,
+        post_torso,
+        action_dim,
+        cfg.network.hidden_state_dim,
+    )
+    q_params = q_net.init(q_key, init_hidden_state, init_x)  # epsilon defaults to 0
+    q_target_params = q_net.init(q_key, init_hidden_state, init_x)  # ensure parameters are separate
+
+    # Pack Q network params
+    params = DDQNParams(q_params, q_target_params)
+
+    # Making optimiser and state
+    opt = optax.chain(
+        optax.clip_by_global_norm(cfg.system.max_grad_norm),
+        optax.adam(learning_rate=cfg.system.q_lr, eps=1e-5),
+    )
+    opt_state = opt.init(params.online)
+
+    # Distribute params, opt states and hidden states across all devices
+    params = replicate(params)
+    opt_state = replicate(opt_state)
+    init_hidden_state = replicate(init_hidden_state)
+
+    # Create dummy transition
+    init_acts = env.action_spec().generate_value()  # (A,)
+    init_transition = Transition(
+        obs=init_obs,  # (A, ...)
+        action=init_acts,
+        reward=jnp.zeros((num_agents,), dtype=float),
+        done=jnp.zeros((1,), dtype=bool),  # one flag for all agents
+    )
+
+    # Initialise trajectory buffer
+    rb = fbx.make_trajectory_buffer(
+        # n transitions gives n-1 full data points
+        sample_sequence_length=cfg.system.sample_sequence_length + 1,
+        period=1,  # sample any unique trajectory
+        add_batch_size=cfg.arch.num_envs,
+        sample_batch_size=cfg.system.sample_batch_size,
+        max_length_time_axis=cfg.system.buffer_size,
+        min_length_time_axis=cfg.system.min_buffer_size,
+    )
+    buffer_state = rb.init(init_transition)
+    buffer_state = replicate(buffer_state)
+
+    # Keys to reset env
+    n_keys = cfg.arch.num_envs * cfg.arch.n_devices * cfg.system.update_batch_size
+    key_shape = (cfg.arch.n_devices, cfg.system.update_batch_size, cfg.arch.num_envs, -1)
+    key, reset_key = jax.random.split(key)
+    reset_keys = jax.random.split(reset_key, n_keys)
+    reset_keys = jnp.reshape(reset_keys, key_shape)
+
+    # Get initial state and timestep per-device
+    env_state, first_timestep = jax.pmap(  # devices
+        jax.vmap(  # update_batch_size
+            jax.vmap(env.reset),  # num_envs
+            axis_name="batch",
+        ),
+        axis_name="device",
+    )(reset_keys)
+    first_obs = first_timestep.observation
+    first_done = first_timestep.last()[..., jnp.newaxis]
+
+    # Initialise env steps and training steps
+    t0 = jnp.zeros((cfg.arch.n_devices, cfg.system.update_batch_size), dtype=int)
```

Comment:
if we're being specific about train we should also be about act
```suggestion
    t0_act = jnp.zeros((cfg.arch.n_devices, cfg.system.update_batch_size), dtype=int)
```